versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/KeyListBuildState.java

```diff
@@ -15,114 +15,156 @@
-        // CommitLogEntry.keyList still has room
-        addToEmbedded(keyListEntry, entrySize);
-      } else {
-        // CommitLogEntry.keyList is "full", switch to the first KeyListEntity
-        embedded = false;
-        newKeyListEntity();
-        addToKeyListEntity(keyListEntry, entrySize);
-      }
-    } else {
-      // filling linked key-lists via CommitLogEntry.keyListIds
+  void add(KeyListEntry entry) {
+    entries.add(entry);
+  }
+
+  static int nextPowerOfTwo(final int v) {
+    return 1 << (32 - numberOfLeadingZeros(v - 1));
+  }
 
-      if (currentSize + entrySize > maxKeyListEntitySize) {
-        // current KeyListEntity is "full", switch to a new one
-        finishKeyListEntity();
-        newKeyListEntity();
+  List<KeyListEntity> finish() {
+    // Build open-addressing map
+    int openAddressingEntries = openAddressingBuckets();
+    int openAddressingMask = openAddressingEntries - 1;
+    KeyListEntry[] openAddressingHashMap = new KeyListEntry[openAddressingEntries];
+    for (KeyListEntry entry : entries) {
+      int hash = entry.getKey().hashCode();
+      int bucket = hash & openAddressingMask;
+
+      // add to map
+      for (int i = bucket; ; ) {
+        if (openAddressingHashMap[i] == null) {
+          openAddressingHashMap[i] = entry;
+          break;
+        }
+
+        i++;
+        if (i == openAddressingEntries) {
+          i = 0;
+        }
       }
-      addToKeyListEntity(keyListEntry, entrySize);
     }
-  }
 
-  private void finishKeyListEntity() {
-    Hash id = randomHash();
-    newKeyListEntities.add(KeyListEntity.of(id, currentKeyList.build()));
-    newCommitEntry.addKeyListsIds(id);
-  }
+    // Generate all KeyList objects. The first one for the embedded key-list and all other ones
+    // for key-list-entities. Offsets are kept for key-list-entities, because the embedded
+    // key-list's is always the first one.
 
-  private void newKeyListEntity() {
-    currentSize = 0;
-    currentKeyList = ImmutableKeyList.builder();
-  }
+    List<Integer> offsets = new ArrayList<>();
+    List<KeyList> keyLists = new ArrayList<>();
 
-  private void addToKeyListEntity(KeyListEntry keyListEntry, int keyTypeSize) {
-    currentSize += keyTypeSize;
-    currentKeyList.addKeys(keyListEntry);
-  }
+    int bucketSize = 0;
+    int maxBucketSize = maxEmbeddedKeyListSize;
+    ImmutableKeyList.Builder keyListBuilder = ImmutableKeyList.builder();
+    for (int i = 0; i < openAddressingHashMap.length; i++) {
+      KeyListEntry entry = openAddressingHashMap[i];
 
-  private void addToEmbedded(KeyListEntry keyListEntry, int keyTypeSize) {
-    currentSize += keyTypeSize;
-    embeddedBuilder.addKeys(keyListEntry);
-  }
+      if (entry != null) {
+        int entrySize = serializedEntrySize.applyAsInt(entry);
+
+        if (bucketSize + entrySize > maxBucketSize) {
+          maxBucketSize = maxKeyListEntitySize;
+          offsets.add(i);
```

Comment:
Another idea: Use the full `long` value range for the "virtual" hash table, and have sorted listed in each bucket. We'd set index boundaries based on key size and load factor applied to each bucket (to reserve space for future inserts).

On get, we'd scan the list of offsets (same as now) to arrive at the bucket ID.
On update, we'd add to the bucket where the hash lands until it fills up, in which case we split this and following buckets (but not change to previous buckets).